[GTK][WPE] Use GBM when available for WebGL in the web process

carlosgcampos opened the pull request:
#### f4c73c22aa00368b5ca4f38823edb432c18a38d1
<pre>
[GTK][WPE] Use GBM when available for WebGL in the web process
<a href="https://bugs.webkit.org/show_bug.cgi?id=279122">https://bugs.webkit.org/show_bug.cgi?id=279122</a>

Reviewed by Alejandro G. Castro.

The solution we found to get the internal texture created by ANGLE is
weak, so we should only use it as fallback for systems where DMA-BUF is
not available. This patch adds a new impementation using the new
CoordinatedPlatformLayerBuffer to add a new one for DMABuf buffers. It
also adds a new class GraphicsContextGLTextureMapperGBM that will be
used also for the GPU process implementation.

* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gbm/DMABufBuffer.cpp: Added.
(WebCore::generateID):
(WebCore::DMABufBuffer::DMABufBuffer):
(WebCore::DMABufBuffer::setBuffer):
* Source/WebCore/platform/graphics/gbm/DMABufBuffer.h: Added.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLGBM::prepareForDisplay):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.h:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBMTextureMapper.cpp: Removed.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.cpp: Added.
(WebCore::GraphicsContextGLTextureMapperGBM::create):
(WebCore::GraphicsContextGLTextureMapperGBM::GraphicsContextGLTextureMapperGBM):
(WebCore::GraphicsContextGLTextureMapperGBM::~GraphicsContextGLTextureMapperGBM):
(WebCore::GraphicsContextGLTextureMapperGBM::platformInitialize):
(WebCore::GraphicsContextGLTextureMapperGBM::platformInitializeExtensions):
(WebCore::GraphicsContextGLTextureMapperGBM::createDrawingBuffer const):
(WebCore::GraphicsContextGLTextureMapperGBM::freeDrawingBuffers):
(WebCore::GraphicsContextGLTextureMapperGBM::bindNextDrawingBuffer):
(WebCore::GraphicsContextGLTextureMapperGBM::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLTextureMapperGBM::prepareForDisplay):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.h: Added.
* Source/WebCore/platform/graphics/gbm/GraphicsLayerContentsDisplayDelegateGBM.cpp: Added.
(WebCore::GraphicsLayerContentsDisplayDelegateGBM::GraphicsLayerContentsDisplayDelegateGBM):
(WebCore::GraphicsLayerContentsDisplayDelegateGBM::setDisplayBuffer):
* Source/WebCore/platform/graphics/gbm/GraphicsLayerContentsDisplayDelegateGBM.h: Added.
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::createWebProcessGraphicsContextGL):
(WebCore::GraphicsContextGLTextureMapperANGLE::~GraphicsContextGLTextureMapperANGLE):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerContentsDisplayDelegateTextureMapper.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBuffer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferDMABuf.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferDMABuf::create):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::CoordinatedPlatformLayerBufferDMABuf):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::importDMABuf const):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferDMABuf.h: Added.
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/283253@main">https://commits.webkit.org/283253@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4503b087fb542a42e092e9f20f76b9082e9d655d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65393 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44762 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18007 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69417 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15999 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67511 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52607 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16281 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52508 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11071 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41413 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56596 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33132 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38086 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13972 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14876 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58506 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14311 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71122 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64637 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9345 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13773 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59830 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9377 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60105 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7720 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1374 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/86404 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9956 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40573 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15220 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41649 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42830 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41393 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/gbm/DMABufBuffer.cpp:
```diff
@@ -0,0 +1,56 @@
+DMABufBuffer::~DMABufBuffer() = default;
+
+void DMABufBuffer::setBuffer(std::unique_ptr<CoordinatedPlatformLayerBuffer>&& buffer)
+{
```
Shouldn't we assert/check here that the assigned buffer is compatible with the current `m_attributes`, at least in debug builds?

carlosgcampos replied:
It's not really possible, but I see how this api can be confusing. The initial idea was to add an import method to DMABufBuffer instead of exposing the internal buffer. But then I thought that in the future we might want to import a buffer from ANGLE code, so epoxy impl was not desireable. That's why I left the import code in CoordinatedPlatformLayerBufferDMABuf where we always want to use epoxy because we are in the compositor thread. Anyway, I think I can change this API to avoid ther explicit setter.

carlosgcampos replied:
Ah!, I remember now, also because we want to create the buffer using the texture mapper bitmap texture pool, and DMABufBuffer doesn't have access to the TextureMapper.

carlosgcampos replied:
Not that easy in the end, I'm going to leave it as is for now, I'll try to find a better solution once new platform buffers are adopted everywhere.

aperezdc replied:
Ah, I see it's tricky. Let's leave it as-is for now, then.

alexgcastro reviewed (approved):
LGTM! Great patch.

carlosgcampos committed:
[GTK][WPE] Use GBM when available for WebGL in the web process
https://bugs.webkit.org/show_bug.cgi?id=279122

Reviewed by Alejandro G. Castro.

The solution we found to get the internal texture created by ANGLE is
weak, so we should only use it as fallback for systems where DMA-BUF is
not available. This patch adds a new impementation using the new
CoordinatedPlatformLayerBuffer to add a new one for DMABuf buffers. It
also adds a new class GraphicsContextGLTextureMapperGBM that will be
used also for the GPU process implementation.

* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gbm/DMABufBuffer.cpp: Added.
(WebCore::generateID):
(WebCore::DMABufBuffer::DMABufBuffer):
(WebCore::DMABufBuffer::setBuffer):
* Source/WebCore/platform/graphics/gbm/DMABufBuffer.h: Added.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLGBM::prepareForDisplay):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.h:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBMTextureMapper.cpp: Removed.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.cpp: Added.
(WebCore::GraphicsContextGLTextureMapperGBM::create):
(WebCore::GraphicsContextGLTextureMapperGBM::GraphicsContextGLTextureMapperGBM):
(WebCore::GraphicsContextGLTextureMapperGBM::~GraphicsContextGLTextureMapperGBM):
(WebCore::GraphicsContextGLTextureMapperGBM::platformInitialize):
(WebCore::GraphicsContextGLTextureMapperGBM::platformInitializeExtensions):
(WebCore::GraphicsContextGLTextureMapperGBM::createDrawingBuffer const):
(WebCore::GraphicsContextGLTextureMapperGBM::freeDrawingBuffers):
(WebCore::GraphicsContextGLTextureMapperGBM::bindNextDrawingBuffer):
(WebCore::GraphicsContextGLTextureMapperGBM::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLTextureMapperGBM::prepareForDisplay):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.h: Added.
* Source/WebCore/platform/graphics/gbm/GraphicsLayerContentsDisplayDelegateGBM.cpp: Added.
(WebCore::GraphicsLayerContentsDisplayDelegateGBM::GraphicsLayerContentsDisplayDelegateGBM):
(WebCore::GraphicsLayerContentsDisplayDelegateGBM::setDisplayBuffer):
* Source/WebCore/platform/graphics/gbm/GraphicsLayerContentsDisplayDelegateGBM.h: Added.
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::createWebProcessGraphicsContextGL):
(WebCore::GraphicsContextGLTextureMapperANGLE::~GraphicsContextGLTextureMapperANGLE):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerContentsDisplayDelegateTextureMapper.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBuffer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferDMABuf.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferDMABuf::create):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::CoordinatedPlatformLayerBufferDMABuf):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::importDMABuf const):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferDMABuf.h: Added.
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/283253@main
